internal/powerplatform/services/data_record/api_data_record.go

```diff
@@ -0,0 +1,432 @@
+		item := list.(map[string]interface{})
+		if item["ReferencingEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencedEntity"].(string)
+			break
+		}
+		if item["ReferencedEntityNavigationPropertyName"] == relationLogicalName {
+			tableName = item["ReferencingEntity"].(string)
+			break
+		}
+	}
+
+	manyToMany, _ := mapResponse["ManyToManyRelationships"].([]interface{})
+	for _, list := range manyToMany {
+		item := list.(map[string]interface{})
+		if item["Entity1NavigationPropertyName"] == relationLogicalName {
+			tableName = item["Entity1LogicalName"].(string)
+			break
+		}
+		if item["Entity2NavigationPropertyName"] == relationLogicalName {
+			tableName = item["Entity2LogicalName"].(string)
+			break
+		}
+	}
+
+	return tableName
+}
+
+func (client *DataRecordClient) ApplyDataRecord(ctx context.Context, recordId string, environmentId string, tableName string, columns map[string]interface{}) (*DataRecordDto, error) {
+	result := DataRecordDto{}
+
+	environmentUrl, err := client.GetEnvironmentUrlById(ctx, environmentId)
+	if err != nil {
+		return nil, err
+	}
+
+	relations := make(map[string]interface{}, 0)
+
+	for key, value := range columns {
+		if nestedMap, ok := value.(map[string]interface{}); ok {
+			delete(columns, key)
+			if len(nestedMap) > 0 {
+				tableLogicalName, ok := nestedMap["table_logical_name"].(string)
+				if !ok {
+					return nil, fmt.Errorf("table_logical_name field is missing or not a string")
+				}
+				dataRecordId, ok := nestedMap["data_record_id"].(string)
+				if !ok {
+					return nil, fmt.Errorf("data_record_id field is missing or not a string")
+				}
+
+				entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableLogicalName)
+
+				columns[fmt.Sprintf("%s@odata.bind", key)] = fmt.Sprintf("%s/api/data/%s/%s(%s)", environmentUrl, constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName, dataRecordId)
+			}
+		}
+		if nestedMapList, ok := value.([]interface{}); ok {
+			delete(columns, key)
+			relations[key] = nestedMapList
+		}
+	}
+
+	entityDefinition := getEntityDefinition(ctx, client, environmentUrl, tableName)
+
+	method := "POST"
+	apiPath := fmt.Sprintf("/api/data/%s/%s", constants.DATAVERSE_API_VERSION, entityDefinition.LogicalCollectionName)
+
+	if val, ok := columns[entityDefinition.PrimaryIDAttribute]; ok {
+		method = "PATCH"
+		apiPath = fmt.Sprintf("%s(%s)", apiPath, val)
+	} else if recordId != "" {
+		method = "PATCH"
+		apiPath = fmt.Sprintf("%s(%s)", apiPath, recordId)
+	}
+
+	e, _ := url.Parse(environmentUrl)
+	apiUrl := &url.URL{
+		Scheme: e.Scheme,
+		Host:   e.Host,
+		Path:   apiPath,
+	}
+
+	response, err := client.Api.Execute(ctx, method, apiUrl.String(), nil, columns, []int{http.StatusOK, http.StatusNoContent}, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(response.BodyAsBytes) != 0 {
+		json.Unmarshal(response.BodyAsBytes, &result)
+	} else if response.Response.Header.Get("OData-EntityId") != "" {
```

Comment:
`OData-EntityId` could be a cost together with other header values if any

Reply:
I'm not sure if I understand this comment correctly? what you mean by that?